app/ts/utils/SignalEffectWithAbortOnChange.ts

```diff
@@ -0,0 +1,43 @@
+import { Signal } from '@preact/signals'
+import { OptionalSignal } from './OptionalSignal.js'
+import { useEffect } from 'preact/hooks'
+
+type UnionSignalType<T> = T extends OptionalSignal<infer U> ? U | undefined : T extends Signal<infer U> ? U extends Signal<infer Inner> ? Inner : U : never
+type SignalTypes<T extends readonly [...(Signal<unknown> | OptionalSignal<unknown>)[]]> = { [K in keyof T]: UnionSignalType<T[K]> }
+const isOptionalSignal = <T>(signal: Signal<T> | OptionalSignal<T>): signal is OptionalSignal<T> => signal instanceof OptionalSignal
+export function useSignalEffectWithAbortOnChange<T extends readonly (Signal<unknown> | OptionalSignal<unknown>)[]>(signals: T, func: (abortSignal: AbortSignal, ...values: SignalTypes<T>) => Promise<void>, onError: (error: unknown) => void) {
+	return useEffect(() => {
+		let abortController = new AbortController()
+		const cleanupFns: (() => void)[] = []
+		const run = () => {
+			abortController.abort()
+			abortController = new AbortController()
+			const values = signals.map(signal => isOptionalSignal(signal) ? signal.deepValue : signal.value) as SignalTypes<T>
+			func(abortController.signal, ...values).catch(onError)
+		}
+		const subscribeTo = (signal: Signal<unknown>, onChange: () => void) => cleanupFns.push(signal.subscribe(onChange))
+		const setupInner = (optional: OptionalSignal<any>) => {
+			const inner = optional.peek()
+			if (inner !== undefined) subscribeTo(inner, run)
+		}
```

Comment:
We also need to unsubscribe to the inner if the outer signal value changes from defined to undefined.
Also, no `any`.
```suggestion
		const setupInner = (optional: OptionalSignal<unknown>) => {
			const inner = optional.peek()
			if (inner === undefined) /*TODO: unsubscribe inner*/
			else subscribeTo(inner, run)
		}
```

Reply:
the unsubs are done at the end:
```ts
return () => {
			cleanupFns.forEach(fn => fn())
			abortController.abort()
		}
		```

Reply:
That is too late.  You need to unsubscribe from the inner signal anytime the outer signal changes to undefined.  You'll then resubscribe to the inner signal when the outer signal changes back to defined.  You already have the resubscribe setup correctly, but without the unsubscribe you'll end up subscribing many times.  You'll also incorrectly receive update notifications when nothing has changed.